packages/babel-core/src/parse.js

```diff
@@ -6,17 +6,79 @@ import normalizeOptions from "./transformation/normalize-opts";
 
 type AstRoot = BabelNodeFile | BabelNodeProgram;
 
-export default function parse(
+export type ParseResult = AstRoot;
+
+export type FileParseCallback = {
+  (Error, null): any,
+  (null, ParseResult | null): any,
+};
+
+type Parse = {
+  (code: string, callback: FileParseCallback): void,
+  (code: string, opts: ?InputOptions, callback: FileParseCallback): void,
+
+  // Here for backward-compatibility. Ideally use ".parseSync" if you want
+  // a synchronous API.
+  (code: string, opts: ?InputOptions): ParseResult | null,
+};
+
+export const parse: Parse = (function parse(code, opts, callback) {
+  if (typeof opts === "function") {
+    opts = undefined;
+    callback = opts;
+  }
+
+  // For backward-compat with Babel 7's early betas, we allow sync parsing when
+  // no callback is given. Will be dropped in some future Babel major version.
+  if (callback === undefined) return parseSync(code, opts);
```

Comment:
Q: If we wanted to introduce a warning, would it be a breaking change? Or we could do it in a minor/patch version?

Reply:
I mostly added this to make lives easier for people, I actually thought about removing it in this PR, but since it's so trivial leaving it to 8.x isn't really that much of a problem.